x/gov/proposal.go

```diff
@@ -0,0 +1,180 @@
+package gov
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/gov/types"
+)
+
+// SubmitProposal create new proposal given a content
+func (keeper Keeper) SubmitProposal(ctx sdk.Context, content Content) (Proposal, sdk.Error) {
+	if !keeper.router.HasRoute(content.ProposalRoute()) {
+		return Proposal{}, ErrNoProposalHandlerExists(keeper.codespace, content)
+	}
+
+	// Execute the proposal content in a cache-wrapped context to validate the
+	// actual parameter changes before the proposal proceeds through the
+	// governance process. State is not persisted.
+	cacheCtx, _ := ctx.CacheContext()
+	handler := keeper.router.GetRoute(content.ProposalRoute())
+	if err := handler(cacheCtx, content); err != nil {
+		return Proposal{}, ErrInvalidProposalContent(keeper.codespace, err.Result().Log)
+	}
+
+	proposalID, err := keeper.getNewProposalID(ctx)
```

Comment:
I'm not 100% sure about this workflow, as far as I've seen:
- getNewProposalID gets and set into the store the next ProposalID

I would set it after `keeper.SetProposal(ctx, proposal)`